cmd/soroban-rpc/internal/db/event.go

```diff
@@ -0,0 +1,336 @@
+// The events are returned in sorted ascending Cursor order.
+// If f returns false, the scan terminates early (f will not be applied on
+// remaining events in the range).
+//
+//nolint:funlen
+func (eventHandler *eventHandler) GetEvents(
+	ctx context.Context,
+	cursorRange CursorRange,
+	contractIDs [][]byte,
+	topics [][][]byte,
+	eventTypes []int,
+	f ScanFunction,
+) error {
+	start := time.Now()
+
+	rowQ := sq.
+		Select(" id", "event_data", "transaction_hash", "ledger_close_time").
+		From(eventTableName).
+		Where(sq.GtOrEq{"id": cursorRange.Start.String()}).
+		Where(sq.Lt{"id": cursorRange.End.String()}).
+		OrderBy("id ASC")
+
+	if len(contractIDs) > 0 {
+		rowQ = rowQ.Where(sq.Eq{"contract_id": contractIDs})
+	}
+	if len(eventTypes) > 0 {
+		rowQ = rowQ.Where(sq.Eq{"event_type": eventTypes})
+	}
+
+	if len(topics) > 0 {
+		var orConditions sq.Or
+		for i, topic := range topics {
+			if topic == nil {
+				continue
+			}
+			orConditions = append(orConditions, sq.Eq{fmt.Sprintf("topic%d", i+1): topic})
+		}
+		if len(orConditions) > 0 {
+			rowQ = rowQ.Where(orConditions)
+		}
+	}
+
+	rows, err := eventHandler.db.Query(ctx, rowQ)
+	if err != nil {
+		return fmt.Errorf(
+			"db read failed for start ledger cursor= %v end ledger cursor= %v "+
+				"contractIDs= %v eventTypes= %v topics= %v   : %w",
+			cursorRange.Start.String(),
+			cursorRange.End.String(),
+			contractIDs,
+			eventTypes,
+			topics,
+			err)
+	}
+
+	defer rows.Close()
+
+	foundRows := false
+	for rows.Next() {
+		foundRows = true
+		var row struct {
+			eventCursorID   string `db:"id"`
+			eventData       []byte `db:"event_data"`
+			transactionHash []byte `db:"transaction_hash"`
+			ledgerCloseTime int64  `db:"ledger_close_time"`
+		}
+
+		err = rows.Scan(&row.eventCursorID, &row.eventData, &row.transactionHash, &row.ledgerCloseTime)
+		if err != nil {
+			return fmt.Errorf("failed to scan row: %w", err)
+		}
+
+		id, eventData, ledgerCloseTime := row.eventCursorID, row.eventData, row.ledgerCloseTime
+		transactionHash := row.transactionHash
+		cur, err := ParseCursor(id)
+		if err != nil {
+			return fmt.Errorf("failed to parse cursor: %w", err)
+		}
+
+		var eventXDR xdr.DiagnosticEvent
+		err = xdr.SafeUnmarshal(eventData, &eventXDR)
+		if err != nil {
+			return fmt.Errorf("failed to decode event: %w", err)
+		}
+		txHash := xdr.Hash(transactionHash)
+		if !f(eventXDR, cur, ledgerCloseTime, &txHash) {
+			return nil
+		}
+	}
+	if !foundRows {
+		eventHandler.log.
+			WithField("duration", time.Since(start)).
+			WithField("start", cursorRange.Start.String()).
+			WithField("end", cursorRange.End.String()).
+			WithField("contracts", contractIDs).
+			WithField("eventTypes", eventTypes).
+			WithField("Topics", topics).
+			Debugf(
+				"No events found for ledger range: duration= %v start ledger cursor= %v - end ledger cursor= %v contractIDs= %v",
+				time.Since(start),
+				cursorRange.Start.String(),
+				cursorRange.End.String(),
+				contractIDs,
+			)
```

Comment:
nit: no need to double-log fields
```suggestion
			Debugf("No events found for ledger range")
```